app/controllers/departments_controller.rb

```diff
@@ -8,7 +8,7 @@ class DepartmentsController < ApplicationController
   def index; end
 
   def show
-    @pagy, @users = pagy @department.users
+    @pagy, @users = pagy @department.users.includes([:avatar_attachment])
```

Comment:
phân biệt includes eger_load vs preload

Reply:
n+1 query xảy ra khi nào?
nhận biết ntn

Reply:
- **Preload**: Thực hiện từng câu query riêng biệt để lấy data từ DB.
- **Eager Load**: Sử dụng 1 câu query lớn với LEFT JOIN để liên kết nhiều tables. Sau đó query theo điều kiện của bảng references.
- Như trình bày ở trên, Rails có hai cách để tránh n+1 query. Một là tạo query lớn(JOIN các bảng) dựa trên assosication, cách hai là tạo mỗi truy vấn cho mỗi asosication.

  Với method **includes**, Rails sẽ quyết định sẽ sử dụng phương pháp nào là phù hợp nhất. Mặc định nó sẽ sử dụng preload tuy nhiên nếu nó nhận thấy rằng bạn đang sử dụng các điều kiện tìm kiếm trong query hoặc sắp xếp thứ tự các record, nó sẽ gọi tới eager load (Vì những điều này chỉ hoạt động bằng cách JOIN các bảng này với nhau)
 Tuy hiệu quả nhưng không phải lúc nào Rails cũng đúng với lựa chọn của nó hoặc có một lý do nào mà bạn muốn chỉ định một cách cụ thể để query thì có thể chon preload hoặc eager_load. Còn không, có vẻ includes là lựa chọn phù hợp nhất.
-------------------------------------------------------------------------------------
**Kết quả của User.preload(:departments):**
1) User Load (0.9ms)  SELECT `users`.* FROM `users` LIMIT 11
2) Relationship Load (1.2ms)  SELECT `relationships`.* FROM `relationships` WHERE `relationships`.`user_id` IN (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11)
3) Department Load (0.9ms)  SELECT `departments`.* FROM `departments` WHERE `departments`.`id` IN (1, 2, 101)
-------------------------------------------------------------------------------------
**Kết quả của User.eager_load(:departments):**
SQL (0.4ms)  SELECT DISTINCT `users`.`id` FROM `users` LEFT OUTER JOIN `relationships` ON `relationships`.`user_id` = `users`.`id` LEFT OUTER JOIN `departments` ON `departments`.`id` = `relationships`.`department_id` LIMIT 11
  SQL (0.3ms)  SELECT `users`.`id` AS t0_r0, `users`.`email` AS t0_r1, `users`.`phone` AS t0_r2, `users`.`full_name` AS t0_r3, `users`.`password_digest` AS t0_r4, `users`.`remember_digest` AS t0_r5, `users`.`activation_digest` AS t0_r6, `users`.`activation` AS t0_r7, `users`.`role` AS t0_r8, `users`.`activation_at` AS t0_r9, `users`.`reset_digest` AS t0_r10, `users`.`reset_sent_at` AS t0_r11, `users`.`created_at` AS t0_r12, `users`.`updated_at` AS t0_r13, `departments`.`id` AS t1_r0, `departments`.`name` AS t1_r1, `departments`.`description` AS t1_r2, `departments`.`created_at` AS t1_r3, `departments`.`updated_at` AS t1_r4 FROM `users` LEFT OUTER JOIN `relationships` ON `relationships`.`user_id` = `users`.`id` LEFT OUTER JOIN `departments` ON `departments`.`id` = `relationships`.`department_id` WHERE `users`.`id` IN (11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 2)

Reply:
n+1 query xảy ra khi truy vấn xử lí dữ liệu không hiệu quả, thường là việc liên tục query trong các vòng lặp
Dấu hiệu phát hiên n+1 query:

- Trong rails có thể dùng gem bullet để hỗ trợ.
- App chạy chậm do liên tục query
- Dựa trên Query Log